### PR TITLE
LNK-2866: Implement code fix for handling FHIR references expressed as absolute URLs

### PR DESCRIPTION
**Pull request by smailliwcs**
## Summary by CodeRabbit

- **New Features**
	- Introduced the `ReferenceRelativizer` class for enhanced handling of FHIR references, allowing for flexible relativization based on base URLs.
	- Added the `RelativizeReferences` class to process FHIR data and relativize references in report generation contexts.

- **Bug Fixes**
	- Improved the logic in the patient data loading process to ensure proper variable scoping and reduce redundant data storage calls.

- **Tests**
	- Added a test suite for `ReferenceRelativizer` to validate the functionality of the relativization process under various scenarios.

### CODE REVIEW

File: core/src/main/java/com/lantanagroup/link/events/RelativizeReferences.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import java.util.List;
+
+public class RelativizeReferences implements IReportGenerationDataEvent {
+  @Override
+  public void execute(TenantService tenantService, Bundle data, ReportCriteria criteria, ReportContext context, ReportContext.MeasureContext measureContext) {
+    ReferenceRelativizer relativizer = new ReferenceRelativizer(FhirContextProvider.getFhirContext());
+    String baseUrl = tenantService.getConfig().getFhirQuery().getFhirServerBase();
+    relativizer.relativize(data, baseUrl);
+  }
```
[the remark this answers is not in the document]
Comment (smailliwcs) [reply core/src/main/java/com/lantanagroup/link/events/RelativizeReferences.java:20]: Our convention with these event handlers is to let exceptions propagate to the caller (i.e., wherever `EventService.,trigger[Data]Event` is invoked).  For `RelativizeReferences`, which will typically be executed as part of the `AfterPatientDataQuery` event, an unhandled exception will bubble up to `PatientScoop.storePatientData`, where we'll lose one patient's worth of data.

File: core/src/main/java/com/lantanagroup/link/ReferenceRelativizer.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+              .filter(Objects::nonNull)
+              .collect(Collectors.toList());
+    } else {
+      resources = List.of(resource);
+    }
+    FhirTerser fhirTerser = fhirContext.newTerser();
+    return resources.stream()
+            .flatMap(_resource -> fhirTerser.getAllPopulatedChildElementsOfType(_resource, Reference.class).stream())
+            .collect(Collectors.toList());
+  }
```
[the remark this answers is not in the document]
Comment (smailliwcs) [reply core/src/main/java/com/lantanagroup/link/ReferenceRelativizer.java:37]: The call to `FhirTerser.getAllPopulatedChildElementsOfType` should not return null `Reference` elements.  We proceed to call `Reference.getReference` on those elements, which _can_ return null, but we explicitly check for that case (via `StringUtils.isEmpty`).

File: core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            .map(Reference::getReference)
+            .toArray(String[]::new);
+    assertArrayEquals(expectedReferences, actualReferences);
+  }
+
+  @Test
+  public void testRelativizeWithNonNullExpectedBaseUrl() {
+    referenceRelativizer.relativize(list, "https://foo/fhir");
+    doTest("Patient/1", "Patient/2", "https://bar/fhir/Patient/3");
+  }
```
[the remark this answers is not in the document]
Comment (smailliwcs) [reply core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java]: Added some additional test cases.

File: core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+  @Test
+  public void testRelativizeWithNonNullExpectedBaseUrlAndTrailingSlash() {
+    referenceRelativizer.relativize(list, "https://foo/fhir/");
+    doTest("Patient/1", "Patient/2", "https://bar/fhir/Patient/3");
+  }
```
[the remark this answers is not in the document]
Comment (smailliwcs) [reply core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java]: Added some additional test cases.

File: core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  @Test
+  public void testRelativizeWithNullExpectedBaseUrl() {
+    referenceRelativizer.relativize(list, null);
+    doTest("Patient/1", "Patient/2", "Patient/3");
+  }
```
[the remark this answers is not in the document]
Comment (smailliwcs) [reply core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java]: Added some additional test cases.

File: core/src/main/java/com/lantanagroup/link/ReferenceRelativizer.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      if (referenceBaseUrl == null) {
+        continue;
+      }
+      if (cleanExpectedBaseUrl == null || StringUtils.equals(referenceBaseUrl, cleanExpectedBaseUrl)) {
+        reference.addExtension()
+                .setUrl(Constants.OriginalElementValueExtension)
+                .setValue(new StringType(referenceElement));
+        reference.setReference(referenceId.toUnqualified().getValue());
+      }
+    }
```
[the remark this answers is not in the document]
Comment (smailliwcs) [reply core/src/main/java/com/lantanagroup/link/ReferenceRelativizer.java:62]: No additional error handling required.  Added some additional test cases to verify this.

File: core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    referenceRelativizer = new ReferenceRelativizer(FhirContextProvider.getFhirContext());
+  }
+
+  @Before
+  public void initializeList() {
+    list = new ListResource();
+    for (String reference : references) {
+      list.addEntry().getItem().setReference(reference);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (smailliwcs) [reply core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java:34]: The newly added test cases (in the `references` field) include invalid and null URLs.

File: core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (smailliwcs) [reply core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java]: The newly added test cases (in the references field) include invalid and null URLs.

File: core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (smailliwcs) [reply core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java]: The test cases are defined in the `references` field.

File: core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (smailliwcs) [reply core/src/test/java/com/lantanagroup/link/ReferenceRelativizerTests.java:34]: I added an invalid URL (containing non-ASCII characters in the host) and a null URL as part of `e9c314e`.

**Review by ajagann1 — APPROVED** [review]
Looks good to me

**Pull request merged**